### PR TITLE
Update intra-monorepo deps when bumping package version

### PR DESCRIPTION
**Pull request by gravityrail**
When the package version linter detects you need to bump a package's version, it recommends that you run something like `tools/project-version.sh -u 0.3.1-alpha js-packages/components`.

This does not update the versions of any other packages using that package. Yet, such a script exists! It's `tools/check-intra-monorepo-deps.sh -u`!

We don't always want to update other deps, so this patch adds an optional extra `-f` argument to `tools/check-intra-monorepo-deps.sh` and only recommends it when you are manually bumping the project version from a linter error. This preserves existing behaviour in all the places that we call this from a script.

#### Changes proposed in this Pull Request:
* Add optional parameter to tools/project-version.sh which will automatically check and fix intra-monorepo dependencies.

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:

Check out a new branch from this one:

```
git checkout -b test/update-intra-monorepo-deps
```

Bump a JS package:

```
tools/project-version.sh -f -u 0.3.1-alpha js-packages/components
```

Check that all the references to `@automattic/jetpack-components` have been updated, as well as the package itself.

Bump a PHP package:

```
tools/project-version.sh -f -u 1.17.0 packages/tracking
```

Check that all the references to `automattic/jetpack-tracking` have been updated, as well as the package itself.

### CODE REVIEW

**Comment by gravityrail** [comment]
Closing in favor of #20752

**Comment by gravityrail** [comment]
Mistakenly closed this when I didn't mean to.

File: tools/project-version.sh [hunk 1]
```diff
@@ -227,4 +231,10 @@ while IFS=" " read -r C F; do
 	sedver "$BASE/projects/$SLUG/$F" "$PAT" "$VERSION" "version constant $C"
 done < <(jq -r '.extra["version-constants"] // {} | to_entries | .[] | .key + " " + .value' "$FILE")
 
+# Update other dependencies
+
+if ! $SKIP_INTRA_MONOREPO_DEPS; then
+	"$BASE/tools/check-intra-monorepo-deps.sh" "-u"
```
Comment (anomiex) [inline tools/project-version.sh]: You should only do this if a version update actually happened. Not when `-c` was specified, and ideally not if all the version numbers were already up to date.

Also, a nitpick: There's no need to quote the "-u", there's no variable expansion involved in it.

```suggestion
	"$BASE/tools/check-intra-monorepo-deps.sh" -u
```

File: .github/workflows/linting.yml [hunk 2]
```diff
@@ -761,10 +761,16 @@ jobs:
           extensions: mysql, imagick
           coverage: none
 
+      - name: Setup Node
+        uses: actions/setup-node@v2
+        with:
+          node-version: ${{ steps.versions.outputs.node-version }}
```
Comment (anomiex) [inline .github/workflows/linting.yml:767]: What's going on with all these additions? That suggests to me that your change is running extra code that it probably shouldn't.

**Review by anomiex — COMMENTED** [review]
`tools/check-intra-monorepo-deps.sh -u` calls `tools/project-version.sh -u`, which you've now made call `tools/check-intra-monorepo-deps.sh -u`. That seems potentially problematic.

In general I like the fact that `tools/project-version.sh -u` does one job: it updates the specified project's version numbers. If we're going to add this "and then update all the other deps too" behavior, I think I'd rather we have an option to enable that instead of having it happen by default with an option to skip it.

File: .github/workflows/linting.yml (hunk 2 above)
Comment (gravityrail) [reply .github/workflows/linting.yml:767]: I was getting test failures that `pnpx` was missing. I can't remember now what those were - I will try removing these changes to see what breaks.

Comment (gravityrail) [reply .github/workflows/linting.yml]: I suspect that you're right, it was running unnecessary code, and now I made some other changes `pnpx` might not be necessary.

**Comment by gravityrail** [comment]
@anomiex wrote:

> In general I like the fact that tools/project-version.sh -u does one job: it updates the specified project's version numbers. If we're going to add this "and then update all the other deps too" behavior, I think I'd rather we have an option to enable that instead of having it happen by default with an option to skip it.

I am sympathetic to that point of view, but another way of look at it was that `tools/project-version.sh -u` was doing half a job, and this change makes it do a whole job ;)

The question I would ask is: How often would you update a package's version number and NOT update all the other packages that refer to it at the same time? You are effectively breaking the tree, and 100% of the time (unless the package isn't required anywhere) you would need to update the inter-package deps if you want to.

In this patch, you can specify the `-s` option to project-version.sh if you _don't_ want to automatically update those deps. I think ultimately this approach will save more time and confusion (particularly for new devs) but I could be convinced otherwise. That was my reasoning for these defaults, anyway.

**Comment by anomiex** [comment]
> How often would you update a package's version number and NOT update all the other packages that refer to it at the same time?

Let's look at the calls:

* `tools/plugin-version.sh` doesn't really need to, because we don't have anything depending on plugins. I suppose maybe in the future that might change, but for now it'd be wasted time.
* If you wind up with multiple notices from CI about needing to run it, it would be more efficient to do all the updates first then update all the other packages at once.
* `tools/check-intra-monorepo-deps.sh` should not wind up calling itself recursively. That'll just be a huge waste.
* `tools/changelogger-release.sh` you already updated to not update the other packages immediately. The script itself does so later, but only does so for the projects that actually need it instead of blindly doing all of them.
* `.github/files/renovate-helper.sh` wants to do all the updates first and then does one `check-intra-monorepo-deps.sh` at the end.

After all that, it looks like every automated call doesn't want to immediately update all the other packages, and manual calls would only really benefit when there's just a single project needing updating.

**Comment by gravityrail** [comment]
@anomiex those are very good points. So, here's what I just changed:

- allow _optional_ "-f" argument to force-update intra-monorepo dependencies. This preserves existing behaviour in all the places that we call this from a script.
- recommend -f argument when outputting version linting feedback, so that if users copy/paste the command it does the nice automated changelogging and versioning.

Would that work?

**Review by kraftbj — APPROVED** [review]
This worked for me and took the review comments by @anomiex into account.

**Pull request merged**